torch_xla/csrc/helpers.cpp

```diff
@@ -527,11 +526,69 @@ std::pair<xla::XlaOp, xla::XlaOp> XlaHelpers::PromoteSecondValue(
 
 xla::Shape XlaHelpers::GetPromotedShape(const xla::Shape& shape1,
                                         const xla::Shape& shape2) {
-  return xla::ShapeUtil::MakeShape(
-      shape1.element_type(),
-      torch::lazy::GetPromotedShape(
-          runtime::util::ToVector<int64_t>(shape1.dimensions()),
-          runtime::util::ToVector<int64_t>(shape2.dimensions())));
+  std::vector<int64_t> dimensions;
```

Comment:
Should this function be used ever if both shapes are static? Are we re-coding XlaBuilder's implicit broadcasting rules here?

Reply:
Here we are porting the function `torch::lazy::GetPromotedShape` [ref](https://github.com/pytorch/pytorch/blob/7c7208a9e76b887c63fb34e6c4278f849d61aa55/torch/csrc/lazy/core/helpers.cpp#L81)  from pytorch/pytorch/torch code base to pytorch/xla code-base for two reasons:

1.  `torch::lazy::GetPromotedShape` does not have the [xla::Shape::kUnboundedSize](https://github.com/openxla/xla/blob/5592294b1f6c5692a4a9acbb35c8c8be979fa948/xla/shape.h#L99) sentinel exposed to it. My goal here is to limit the xla specific information in the PyTorch/X:LA code base. 
2. `torch::lazy::GetPromotedShape`  does not properly set the [dynamic_dimension](https://github.com/openxla/xla/blob/5592294b1f6c5692a4a9acbb35c8c8be979fa948/xla/shape.h#L303). The clone has those changes incorporated. 

Note that  `ImplicitBroadcastWithUnboundedDynamicShapes` in any case has to infer final shape of broadcast-in-dim custom calls and the inference logic is very similar to  `torch::lazy::GetPromotedShape` except that the `dynamic_dimension` of the result shape is not set.